salt/modules/aptpkg.py

```diff
@@ -1756,23 +1756,32 @@ def _get_opts(line):
     return ret
 
 
-def _split_repo_str(repo):
+def apt_source_entry(repo):
     """
-    Return APT source entry as a tuple.
+    .. versionadded:: 3006.2
```

Comment:
If this is a new feature this would need to be 3007.0, not 3006.2 and a changelog would also need to be added for adding this new function in addition to the one you current have added. ( a changelog with a `.added.md` extension)

Reply:
But a user could not have previously run this function, and now they can, so that part is a new feature. We could separate the PRs out:
1. push a PR to the `3006.x` branch keeping the private function with the changes, without the `versionadded`.
2. keep this PR pushed to master, but updating the versionadded to 3007.0.

Reply:
We could import the private function and namespace it into the state module?
That way we would be coherent?

Reply:
@terminalmage I can certainly see your point here. Let me discuss with the team and I'll get back to you here.

Reply:
I do not understand what is incoherent about this PR.

I also do not understand how importing and invoking a private function is more coherent than checking if the function is in the loader, as salt/states/pkgrepo.py already does [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkgrepo.py#L361), [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkgrepo.py#L688), and [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkgrepo.py#L136).

Reply:
That's not the point, the point is exposing(adding) new functionality in a point release, a feature addition.
Yes, the import is more cumbersome, and would have to be undo for the `master` branch to do it your way...

Anyway, let's see what the rest of the team has to say about this.

Reply:
The import should be done after the Debian family check

Reply:
This solution would work, sure, but what about cases where people are manually specifying the provider (via the [providers](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#providers) minion config option)? This is one of the reasons we have historically tried to avoid grains checks in the states, opting instead for `__salt__` membership checks.

Sometimes core grains become out-of-date, when new distros arise (or new versions of distros do things differently), leading to lapses in Salt's ability to properly detect OS. In those cases, `providers` is a good way to get things working while Salt catches up to these changes. The proposed alternative would break `pkgrepo.absent` for these minions.

I can think of 4 solutions (please suggest alternatives if any come to mind):


1. Implement the proposed alternative solution from [this comment](https://github.com/saltstack/salt/pull/64287#discussion_r1212065476).
   - This would break minions which are manually set to use aptpkg as the `pkg` provider, as mentioned above.
2. Implement the proposed alternative solution, but also check `__opts__` to see if `providers` is in use for `pkg`.
   - `__opts__.get('providers', {}).get('pkg') == 'aptpkg'`
   - While this would technically work, IMO it defeats the purpose of using `providers`.
   - Furthermore, it breaks with established precedent of using loader membership to run code that is platform-specific. Some examples:
     1. `pkg.check_db` only exists for Gentoo Linux (and derivatives which use `portage`). The function serves little purpose for CLI usage, and is arguably there simply for programmatic purpose. Yet, `pkg` states [check the loader](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L946-L962) to see whether or not that function is there to be run.
     2. `pkg.list_provides` only exists for SUSE, yet Salt checks the loader [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L565-L574) and [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L2030-L2034) to see whether or not this function is available to be run.
     4. The `lowpkg` virtual module only exists for systems that use DEB or RPM packages. Yet states [check if `lowpkg.bin_pkg_info` is present](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L756) to see whether or not it can parse package metadata for states which use `sources` to install package files without using a DEB/RPM repository. The functions in `lowpkg` don't serve a meaningful purpose on the CLI and are used for programmatic reasons.
     5. `pkg.verify` only exists for `yum` and `zypper`, yet states [check the loader](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L1747) to make sure that package verification is available.
3. Keep the code as-is, but target 3007.0. (docstrings would need to be updated). This decision would be my least favorite, since it would be a conscious decision to 2 allow easily-fixable bugs to remain in the codebase.
4. Make no changes to the pull request.

Reply:
We would like to go forward with option 2. The issue isn't adding the ability to check to see if the function exists in the loader, the issue is we are making a function available to users that does not add value, but only for programmatic purposes. If this is something that is already being done in the code today, we should consider that technical debt that would need to be fixed.

Reply:
This has been completed.